### ISSUE TITLE
NET-1605 Fix S2583 FP: Warns about captured variable changed in local function

### ISSUE THREAD

**Issue by kevinoid** [SonarSource/sonar-dotnet#8885]
### Description

[S2589](https://sonarsource.github.io/rspec/#/rspec/S2589/csharp) warns that a variable which is only assigned in a [local function](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/local-functions) always evaluates to the same value.

### Repro steps

```csharp
using System;
using System.Timers;

public static class Program
{
    public static int Main(string[] args)
    {
        bool elapsed = false;
        void OnElapsed(object? source, ElapsedEventArgs e)
        {
            elapsed = true;
        }

        var timer = new Timer(2000);
        timer.Elapsed += OnElapsed;
        timer.Enabled = true;

        while (!elapsed)
        {
            Console.WriteLine("Timer not elapsed.");
            System.Threading.Thread.Sleep(500);
        }

        Console.WriteLine("Timer elapsed!");
        return 0;
    }
}
```

### Expected behavior

[S2589](https://sonarsource.github.io/rspec/#/rspec/S2589/csharp) is not produced.

### Actual behavior

    /path/to/Program.cs(18,17): warning S2583: Change this condition so that it does not always evaluate to 'False'. Some code paths are unreachable.

### Known workarounds

The warning can be suppressed.

### Related information

* SonarAnalyzer.CSharp version: 9.21.0.86780

**Comment by cristian-ambrosini-sonarsource** [comment]
Hi @kevinoid!
I confirm this as a FP.
At this stage, our Symbolic Execution engine is not tracking events nor delegates, resulting in this kind of false positives.
We don't have a specific timeline for this issue to be fixed but I'll add it to our backlog.
Thanks!

**Comment by jilles-sg** [comment]
@cristian-ambrosini-sonarsource I have disabled this rule in our configuration because I think the risk is too high that someone removes a condition that the rule says is redundant but is actually required. It would be better if the rule does not report anything if the affected code is too hard to analyze properly.

It's sad because this kind of rule would be very useful if it did not cause false positives (some false negatives are acceptable).

By the way, I consider the reproducer to have a threading bug (unsynchronized access to `elapsed`).

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1605](https://sonarsource.atlassian.net/browse/NET-1605)

[NET-1605]: https://sonarsource.atlassian.net/browse/NET-1605?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
